### PR TITLE
[#14282] Add the pinpoint-alarm-batch module

### PR DESCRIPTION
**Pull request by ga-ram**
Follows #14288. That change shipped the alarm model, dao, evaluation and senders, but nothing ran them on a schedule. This module adds that: the job that evaluates due rules, and the dispatcher that drains the notification outbox.

**Scope: library only.** Nothing in this repository imports this module yet, so merging it starts no sweep — that is deliberate, the same shape as #14288, which added the alarm model and dao with no consumer. A distribution that wants scheduled alarms imports `AlarmJobSchedulerConfiguration` and `AlarmNotificationSchedulerConfiguration` alongside its own data source and batch metadata configuration. Wiring it into a deployable here is a separate change, once this repository has a `MetricQueryService` to evaluate rules against.

## The two schedules are separate on purpose

Evaluation records the event and enqueues its deliveries in one transaction; sending them is a network call that can fail, retry, and take as long as the far side takes. Keeping them apart means a failed send cannot roll back the event that caused it, and a slow provider cannot delay the next evaluation sweep.

## No backend knowledge in the job

Rules are routed by the data source code they store, so the module that owns a data source contributes the `MetricQueryService` for it. A rule whose data source no installed module owns is reported to that rule's own channels, because only its owner can fix it — a backend outage is not, because every rule against that backend would say the same thing at once.

## Per-rule timeouts, measured from the wait rather than the queue

Each rule is submitted to the executor and waited on for the whole timeout, one at a time. A deadline attached at submission would expire on rules that never got a thread — the executor runs a few at a time — marking them failed without evaluating them. Since the recorded failure also moves the rule's next check forward, an alarm that should have fired would simply be skipped.

A rule that exceeds its timeout is cancelled, which frees its thread. That is a request rather than a guarantee: a query already blocked in the driver can ignore the interrupt, finish, and commit a state contradicting the failure recorded for it. Closing that window needs a deadline the query itself honours or a generation on the state row, and the comment at that line says so rather than claiming more than the code does.

## A sweep carries a cursor across its continuations

Evaluating a rule normally moves its next check forward, which is what stops the next query returning it. But recording a failure is itself allowed to fail, and without a cursor a batch of rules whose state could not be written comes back on every continuation and the sweep never ends — on a single-threaded scheduler that stops alarms altogether.

## The step runs without a transaction

Everything the tasklet writes goes through its own `REQUIRES_NEW` transactions, so one rule's failure cannot roll back another's, and the job repository writes its metadata to a different data source. A transaction at the step would have nothing to roll back and nothing to make consistent; it would only hold an alarm connection and its read view open for the length of the whole sweep. Spring Batch's own metadata configuration is not shipped here either — this repository already has `MetaDataBatchConfiguration`, and two `DefaultBatchConfiguration` subclasses in one context would define `jobRepository` twice.

## Webhook calls are host-checked at connection time

A webhook URL is typed in by an operator, so the client that calls it enforces the host policy and refuses redirects. The check runs in the DNS resolver rather than before the call, so it sees the addresses the connection will actually use — a host that passed validation at registration cannot be re-pointed at an internal one afterwards. A deployment whose own notification back-end legitimately lives on a private address gives that transport its own client rather than relaxing this one.

## A sender exists only when its transport is configured

Email once a From address is set — blank counts as unset, so a deployment that leaves it out gets no email sender rather than a startup failure on a missing `JavaMailSender`. No SMS sender at all, since a gateway is bought rather than installed. Nothing stands in to accept a notification and drop it: a channel with no sender for its method type is logged and recorded against the alarm's history where the rule's owner can see it, and the rule's other channels still go out. For the same reason the email dispatcher fails the delivery when no address in the group could be parsed, instead of returning quietly and having it recorded as sent.

## Testing

23 integration cases against a real MySQL (Testcontainers), covering what unit tests cannot: the alarm lifecycle across sweeps, action-interval suppression, the transaction boundary between recording an event and enqueueing its deliveries, and outbox lease contention between concurrent dispatchers. They run the shipped job, step and tasklet rather than rebuilding them, so the step's transaction handling and the continuation cursor are part of what is exercised. The schema comes from the shipped DDL rather than a fixture copy, because a copy drifts silently and the tests that would catch a changed column are the ones running against it.

`alarm` 299 tests and `alarm-batch` 33 pass; the full build is green.

## Known follow-ups, not in this PR

- **No retention for `alarm_notification_outbox`.** Terminal rows accumulate. Dispatch reads are protected by `idx_outbox_dispatch`, and cleanup can join `alarm_history_v2.created_at` without a schema change. `AlarmHistoryV2Dao.deleteOlderThan` has no caller yet and belongs in the same change.
- **The evaluation sweep drives from `alarm_rule_v2` and cannot use `idx_alarm_state_due`.** Switching to an `alarm_state`-driven query needs `AlarmRuleService.updateRule` to upsert the state row on an enabled transition first, or rules enabled through it would never be evaluated.
- **The webhook client configuration overlaps `batch-alarmsender`,** down to the bean names, and their timeouts have already drifted. Consolidating them touches both modules.
- **A failure message reaches a webhook unfiltered.** The text a notification carries for a check failure is the exception's root cause message, which is not the exception the classifier matched on, so an arbitrary message can be posted to an operator-supplied URL.

The second commit adds the Apache license header to the `alarm` module's Java sources, which went in without one.

### CODE REVIEW

File: alarm-batch/src/main/java/com/navercorp/pinpoint/alarm/batch/config/AlarmJobConfiguration.java [hunk 1]
```diff
@@ -0,0 +1,243 @@
...
+                try {
+                    future.get(ruleTimeoutSec, TimeUnit.SECONDS);
+                } catch (TimeoutException timeout) {
+                    // Interrupts the run so it stops at the next interruptible point and frees
+                    // its thread. It is not a guarantee: a query already blocked in the driver
+                    // can ignore it, finish, and commit a state that contradicts the failure
+                    // recorded below. Closing that window needs a deadline the query itself
+                    // honours, or a generation on the state row -- neither is here yet.
+                    future.cancel(true);
+                    recordFailure(evaluationService, failureClassifier, rule, timeout);
```
[the remark this answers is not in the document]
Comment (ga-ram) [reply alarm-batch/src/main/java/com/navercorp/pinpoint/alarm/batch/config/AlarmJobConfiguration.java:205]: **Current state**

The overclaim is fixed: `AlarmJobConfiguration.java:199-203` no longer says cancellation prevents a contradicting commit, only that the interrupt frees the thread where the work can see it. No fence was added.

**Why the overwrite is not reachable as configured**

- A late run can only overwrite with a *success*, so the query must return data after the wait gave up. That is bounded by the data source client, not by the cancel — the Pinot client, for one, aborts at `pinot.broker.timeoutMs` (default 10s) and abandons an unanswered socket at `brokerReadTimeoutMs` (default 60s). Beyond those the task throws, and an exception on a future nobody awaits writes nothing.
- The same rule is not re-evaluated until `sweepDuration + fixedDelay` after this sweep started — fixed delay, single-threaded scheduler, single process (`AlarmJobSchedulerConfiguration.java:33-44`). That exceeds the deadline above, so a late write lands before anything newer exists.
- The failure is not lost either way: `recordCheckFailed` (`AlarmEventPersistenceService.java:142-167`) inserts an `alarm_history_v2` row and enqueues its notification, and nothing later removes them. A late success changes `alarm_state.status` only, to what the measurement found.

**What would open the window**

1. More than one evaluating instance — `selectDueEnabledRulesAfter` (`AlarmRuleV2Mapper.xml:250`) takes no claim. Documented at `AlarmJobSchedulerConfiguration.java:36-44`.
2. A data source whose client enforces no deadline shorter than the sweep interval. This module cannot check for that.

**Deferred fix**

Conditional persistence, no new column: `alarm_state.last_checked_at` (`CreateTableStatement-mysql.sql:87`) carries the time the evaluation *started*, and the write updates under `AND (last_checked_at IS NULL OR last_checked_at <= #{evaluationStartedAt})`. That rejects a superseded evaluation and still accepts a late one nothing has overtaken. `DATETIME` resolution suffices, since two evaluations of one rule are at least a sweep interval apart.

Deferred because it threads `evaluationStartedAt` through `AlarmEvaluationService` and four `record*` methods for a window that is currently closed, and because condition 1 needs due-rule claiming anyway. Tracked with that work.

**On the query-deadline half of the suggestion**

Out of scope here. `alarm-batch` ships no `MetricQueryService`; it routes by data source code and the implementations are contributed elsewhere (`AlarmJobConfiguration.java:136,171`). A query deadline belongs on the client that issues the query, configured by the module owning that data source, next to the connection settings it must stay consistent with.

File: alarm-batch/src/main/java/com/navercorp/pinpoint/alarm/batch/config/AlarmNotificationSchedulerConfiguration.java [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+        scheduler.setWaitForTasksToCompleteOnShutdown(true);
+        scheduler.setAwaitTerminationSeconds(60);
+        return scheduler;
+    }
+
+    @Scheduled(
+            fixedDelayString = "${pinpoint.modules.batch.alarm.outbox.fixedDelayMs:5000}",
+            scheduler = "alarmOutboxScheduler")
+    public void dispatchAlarmNotifications() {
+        notificationDispatcher.dispatch();
```
[the remark this answers is not in the document]
Comment (ga-ram) [reply alarm-batch/src/main/java/com/navercorp/pinpoint/alarm/batch/config/AlarmNotificationSchedulerConfiguration.java:59]: **Current state**

Bounded, but not by either of the two suggestions. `AlarmNotificationDispatcher.dispatch()` (`AlarmNotificationDispatcher.java:114-136`) computes `leaseExpiry` from the claim time and checks it before each send; once it has lapsed the run stops and leaves the remaining rows claimable for the next one.

```java
LocalDateTime leaseExpiry = claimTime.plus(leaseDuration);
for (AlarmNotificationOutbox delivery : deliveries) {
    if (!LocalDateTime.now(clock).isBefore(leaseExpiry)) {
        logger.warn("Lease ran out with {} of {} deliveries left; ...");
        break;
    }
    dispatch(delivery);
}
```

With the defaults (`AlarmNotificationDispatcher.java:68-70`: 100 rows, 60s lease) and a 10s webhook response timeout (`AlarmNotificationConfiguration.java:74`), that reduces the rows exposed to a concurrent re-claim from up to 100 to at most one.

**The one that remains**

The check runs before a send, so it does not bound the send already in flight when the lease lapses: a send starting at 59.9s and taking 10s finishes past expiry, and that row can be re-claimed and sent again meanwhile. This is stated in the comment at `AlarmNotificationDispatcher.java:117-122` rather than left implied — an earlier revision claimed the run stays inside its lease, which was wrong.

**Why not lease renewal or a bounded claim**

Renewal makes the lease a heartbeat: it needs a write per send and has to decide what happens when the renewal itself fails mid-batch, which is more machinery than one overlapping row warrants. Bounding `batchSize × maxSendDuration ≤ lease` ties two independently tuned settings to a per-send maximum that only the webhook transport actually has — email has no equivalent bound — so it would be enforcing a number the code cannot derive.

Both stay available if the residual overlap turns out to matter; the dispatch loop is where either would go.
